Thistle/galley-sdk
# FRU-83 Create Bulk Mutation to for adding "Core Recipe" Preparation tags

## Description
This features adds a bulk mutation function to the adds a "Core Recipe" Preparation tag to all BASE Recipe Items provided. The script requires as input, a set of recipe item IDs provided by the Culinary OPS, locates all recipe items in Galley, and for each the base recipe (Usage Recipe) and appends a "Core Recipe" tag to the Preparations field. The feature only appends this tag, if the tag is not previously there.


## Test Plan
For the manual test plan, we want to use the feature to create a fresh new set of base recipe items that with the "Core Recipe" Preparations tag
1. Sign in to Galley [here](https://staging-app.galleysolutions.com/login)
2. Go to [Preparation: Core Recipe](https://staging-app.galleysolutions.com/preparations/cHJlcGFyYXRpb246MzEzNjk=) of the Galley staging-app to view all recipes that contain a "Core Recipe" preparation

### Testing a single recipe id:
3. Open terminal and run `python` from the `galley-sdk` root folder:
```python3
>>> from galley.queries import *; from galley.enums import *; from galley.mutations import *; from galley.management import *;

>>> get_untagged_core_recipe_item_ids(["cmVjaXBlOjIwMjI5NA=="])    # recipe id for Harissa Roasted Carrot Salad BASE
```
- [x]  Verify result: `['cmVjaXBlSXRlbToxMjkwMTgx', 'cmVjaXBlSXRlbToxMjkwMjAw']`

    - Note: These two recipe item ids correspond to the [Harissa Roasted Carrot Salad BASE / "cmVjaXBlOjIwMjI5NA=="](https://staging-app.galleysolutions.com/recipes/cmVjaXBlOjIwMjI5NA==) as a unique component of its two parent recipes (scroll to bottom of the page under the "USAGES" section):
        - (1) Harissa Roasted Carrot Salad with Paprika Chicken
        - (2) Harissa Roasted Carrot Salad with Paprika Kidney Beans

5. Run the batch job for adding Core Recipe preparation to all the recipe items that require the **Harissa Roasted Carrot Salad BASE** recipe above:  
```python3
>>> update_recipe_item_data({"ids": ["cmVjaXBlOjIwMjI5NA=="], "attrs": {"preparationIds": ["cHJlcGFyYXRpb246MzEzNjk="]}})
```
- Note: `update_recipe_item_data()` automatically runs `get_untagged_core_recipe_item_ids()` when called.

6.  Refresh page from Step 2 and verify that these recipes are shown to now contain a "Core Recipe" preparation:
    - [x] (1) Harissa Roasted Carrot Salad with Paprika Chicken
    - [ ] (2) Harissa Roasted Carrot Salad with Paprika Kidney Beans

### Testing multiple recipe ids:
7. Run method with three recipe ids passed in:   
```python3
>>> get_untagged_core_recipe_item_ids(["cmVjaXBlOjIwMjI5NA==", "cmVjaXBlOjE3NjQxNA==", "cmVjaXBlOjE4MDAwMw=="])    
# recipe ids for Harissa Roasted Carrot Salad BASE, Daikon Sunflower Salad BASE, & Roasted Apple & Squash Chipotle Chili BASE
```
- [x] Verify result: `['cmVjaXBlSXRlbToxMTUwNDY5', 'cmVjaXBlSXRlbToxMTUwNDE3', 'cmVjaXBlSXRlbToxMTY0Njg5', 'cmVjaXBlSXRlbToxMTY0NTgy', 'cmVjaXBlSXRlbToxMjkyNjU3', 'cmVjaXBlSXRlbToxMjkyNjYy']`
    
    - Note: The ids from Step 3 should not be included in this return list since they've already had the "Core Recipe" preparation applied in Step 5

9. Run the batch job for adding "Core Recipe" preparation to all recipe items that require the Harissa Roasted Carrot Salad BASE,  Daikon Sunflower Salad BASE, & Roasted Apple & Squash Chipotle Chili BASE. 
```python3
update_recipe_item_data({"ids": ["cmVjaXBlOjIwMjI5NA==", "cmVjaXBlOjE3NjQxNA==", "cmVjaXBlOjE4MDAwMw=="], "attrs": {"preparationIds": ["cHJlcGFyYXRpb246MzEzNjk="]}})
```
10. Refresh page from Step 2 and verify that these recipes are shown to now contain a "Core Recipe" preparation:
    - 2 usage items for [Harissa Roasted Carrot Salad BASE / cmVjaXBlOjIwMjI5NA==](https://staging-app.galleysolutions.com/recipes/cmVjaXBlOjIwMjI5NA==)
        - [x] (1) Harissa Roasted Carrot Salad with Paprika Chicken
        - [x] (2) Harissa Roasted Carrot Salad with Paprika Kidney Beans
    - 2 usage items for [Daikon Sunflower Salad BASE / cmVjaXBlOjE3NjQxNA==](https://staging-app.galleysolutions.com/recipes/cmVjaXBlOjE3NjQxNA==)
        - [x] (1) Daikon Sunflower Salad with Ginger Maple Dressing and Garlic Chicken
        - [x] (2) Daikon Sunflower Salad with Ginger Maple Dressing and Garlic Mung Beans
    - 4 usage items for [Roasted Apple & Squash Chipotle Chili BASE / cmVjaXBlOjE4MDAwMw==](https://staging-app.galleysolutions.com/recipes/cmVjaXBlOjE4MDAwMw==)
        - [x] (1) Roasted Apple & Squash Chipotle Chili with Herb Pinto Beans
        - [x] (2) Roasted Apple & Squash Chipotle Chili with Paprika Chicken
        - [x] (3) Roasted Apple & Squash Chipotle Chili with Savory Ground Turkey
        - [x] (4) Roasted Apple & Squash Chipotle Chili with Savory Pinto Beans
11. Notice that we've now ran the command `update_recipe_item_data()` twice on [Harissa Roasted Carrot Salad BASE / cmVjaXBlOjIwMjI5NA==](https://staging-app.galleysolutions.com/recipes/cmVjaXBlOjIwMjI5NA==). Go to parent recipe pages and verify that **ONLY ONE** "Core Recipe" preparation chip is present for the Harissa Roasted Carrot Salad BASE recipe under the "COMPONENTS" section:
    - [x] (1) [Harissa Roasted Carrot Salad with Paprika Chicken](https://staging-app.galleysolutions.com/recipes/cmVjaXBlOjIwMjMwNQ==)
    - [x] (2) [Harissa Roasted Carrot Salad with Paprika Kidney Beans](https://staging-app.galleysolutions.com/recipes/cmVjaXBlOjIwMjMwMA==)

### Automated Test
```python -m unittest tests/test_*```
